TODO: Add Title
TODO: Add TOC
TODO: Add Goals

In this notebook, we will load the cleaned dataset created in the first notebook (full_format_recipes_cleaned.csv) and do preprocessing on the data to get it ready for modeling.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm

from sklearn.feature_extraction.text import CountVectorizer


In [2]:
f"pandas version: {pd.__version__}"

'pandas version: 2.1.4'

## Load the data

In [3]:
df = pd.read_csv('../data/interim/full_recipes_cleaned_2.csv')
df.shape

(14526, 7)

In [4]:
df.head()

recipeId  calories  rating                                       title  \
0         0     426.0   2.500              Lentil, Apple, and Turkey Wrap   
1         1     403.0   4.375  Boudin Blanc Terrine with Red Onion Confit   
2         2     165.0   3.750                Potato and Fennel Soup Hodge   
3         4     547.0   3.125                    Spinach Noodle Casserole   
4         5     948.0   4.375                               The Best Blts   

                                       directionsStr  \
0  ['1. Place the stock, lentils, celery, carrot,...   
1  ['Combine first 9 ingredients in heavy medium ...   
2  ['In a large heavy saucepan cook diced fennel ...   
3  ['Preheat oven to 350°F. Lightly grease 8x8x2-...   
4  ['Mix basil, mayonnaise and butter in processo...   

                                       categoriesStr  \
0  ['Sandwich', 'Bean', 'Fruit', 'Tomato', 'turke...   
1  ['Food Processor', 'Onion', 'Pork', 'Bake', 'B...   
2  ['Soup/Stew', 'Dairy', 'Potato', 'Vegetable', ...   
3  ['Cheese', 'Dairy', 'Pasta', 'Vegetable', 'Sid...   
4  ['Sandwich', 'Food Processor', 'Tomato', 'Kid-...   

                                      ingredientsStr  
0  ['4 cups low-sodium vegetable or chicken stock...  
1  ['1 1/2 cups whipping cream', '2 medium onions...  
2  ['1 fennel bulb (sometimes called anise), stal...  
3  ['1 12-ounce package frozen spinach soufflé, t...  
4  ['2 1/2 cups (lightly packed) fresh basil leav...

In [5]:
df.set_index('recipeId')

calories  rating                                       title  \
recipeId                                                                 
0            426.0   2.500              Lentil, Apple, and Turkey Wrap   
1            403.0   4.375  Boudin Blanc Terrine with Red Onion Confit   
2            165.0   3.750                Potato and Fennel Soup Hodge   
4            547.0   3.125                    Spinach Noodle Casserole   
5            948.0   4.375                               The Best Blts   
...            ...     ...                                         ...   
20125         28.0   3.125                              Parmesan Puffs   
20126        671.0   4.375              Artichoke and Parmesan Risotto   
20127        563.0   4.375                       Turkey Cream Puff Pie   
20128        631.0   4.375     Snapper on Angel Hair with Citrus Cream   
20129        560.0   4.375  Baked Ham with Marmalade-Horseradish Glaze   

                                              directionsStr  \
recipeId                                                      
0         ['1. Place the stock, lentils, celery, carrot,...   
1         ['Combine first 9 ingredients in heavy medium ...   
2         ['In a large heavy saucepan cook diced fennel ...   
4         ['Preheat oven to 350°F. Lightly grease 8x8x2-...   
5         ['Mix basil, mayonnaise and butter in processo...   
...                                                     ...   
20125     ['Beat whites in a bowl with an electric mixer...   
20126     ['Bring broth to simmer in saucepan.Remove fro...   
20127     ['Using a sharp knife, cut a shallow X in bott...   
20128     ['Heat 2 tablespoons oil in heavy medium skill...   
20129     ['Position rack in bottom third of oven and pr...   

                                              categoriesStr  \
recipeId                                                      
0         ['Sandwich', 'Bean', 'Fruit', 'Tomato', 'turke...   
1         ['Food Processor', 'Onion', 'Pork', 'Bake', 'B...   
2         ['Soup/Stew', 'Dairy', 'Potato', 'Vegetable', ...   
4         ['Cheese', 'Dairy', 'Pasta', 'Vegetable', 'Sid...   
5         ['Sandwich', 'Food Processor', 'Tomato', 'Kid-...   
...                                                     ...   
20125     ['Mixer', 'Cheese', 'Egg', 'Fry', 'Cocktail Pa...   
20126     ['Side', 'Kid-Friendly', 'High Fiber', 'Dinner...   
20127     ['Onion', 'Poultry', 'turkey', 'Vegetable', 'B...   
20128     ['Milk/Cream', 'Citrus', 'Dairy', 'Fish', 'Gar...   
20129     ['Pork', 'Bake', 'Roast', 'Christmas', 'Ham', ...   

                                             ingredientsStr  
recipeId                                                     
0         ['4 cups low-sodium vegetable or chicken stock...  
1         ['1 1/2 cups whipping cream', '2 medium onions...  
2         ['1 fennel bulb (sometimes called anise), stal...  
4         ['1 12-ounce package frozen spinach soufflé, t...  
5         ['2 1/2 cups (lightly packed) fresh basil leav...  
...                                                     ...  
20125     ['2 large egg whites', '3 oz Parmigiano-Reggia...  
20126     ['5 1/2 cups (or more) low-salt chicken broth'...  
20127     ['1 small tomato', '1 small onion, finely chop...  
20128     ['4 tablespoons olive oil', '4 shallots, thinl...  
20129     ['1 18-pound fully cooked bone-in smoked ham, ...  

[14526 rows x 6 columns]

In [6]:
# confirm that there are no null values or duplicated values
print(f"Null values: {df.isna().sum().sum()}")
print(f"Duplicated rows: {df.duplicated().sum()}")

Null values: 0
Duplicated rows: 0


In [39]:
measurements = set(line.strip() for line in open('..\..\Capstone\measurement_list.txt'))
extra_adjectives = set(line.strip() for line in open('..\..\Capstone\extra_adjectives_list.txt'))
extra_words = set(line.strip() for line in open('..\..\Capstone\extra_words_list.txt'))

In [40]:
from sklearn.feature_extraction._stop_words import ENGLISH_STOP_WORDS

"""
# one hundred least useful tokens
useless = list(coefs.abs().sort_values()[:100].index)
",".join(useless)

my_stops = set(ENGLISH_STOP_WORDS) | set(useless)
print(len(my_stops))
print(",".join(my_stops))
"""
my_stops = set(ENGLISH_STOP_WORDS) | measurements | extra_adjectives | extra_words

def my_tokenizer(text):
    # convert to lowercase
    text = text.lower()
    # break into characters and weed out punctuation etc.  (include space!)
    chars = list(char for char in text if char in "abcdefghijklmnopqrstuvwxyz ")
    # make back into a single string
    text = "".join(chars)
    # break into words and weed out stop words and short words < 3 characters
    text = list(word for word in text.split() if word not in my_stops and len(word) >=3)
    return text

my_tokenizer(df['ingredientsStr'][1])

['whipping',
 'cream',
 'onions',
 'salt',
 'bay',
 'leaves',
 'cloves',
 'garlic',
 'clove',
 'pepper',
 'nutmeg',
 'thyme',
 'shallots',
 'butter',
 'boneless',
 'center',
 'pork',
 'loin',
 'sinew',
 'cut',
 'chunks',
 'chilled',
 'eggs',
 'purpose',
 'flour',
 'tawny',
 'port',
 'currants',
 'lettuce',
 'leaves',
 'peppercorns',
 'fresh',
 'parsley',
 'bay',
 'leaves',
 'french',
 'bread',
 'baguette',
 'slices',
 'olive',
 'oil',
 'red',
 'onions',
 'halved',
 'currants',
 'red',
 'wine',
 'vinegar',
 'canned',
 'chicken',
 'broth',
 'fresh',
 'thyme',
 'sugar']

In [41]:
vect = CountVectorizer(tokenizer=my_tokenizer,
                       min_df=5)
vect.fit(df['ingredientsStr'])
vocab = vect.get_feature_names_out()
text_transformed = vect.transform(df['ingredientsStr'])

c:\Users\anami\anaconda3\Lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [42]:
len(vocab)

2168

In [43]:
type(vocab)

numpy.ndarray

In [35]:
#vocab.tofile("vocab",sep=" ", format="%s")

In [52]:
vect2 = CountVectorizer(stop_words="english")
vect2.fit(df['categoriesStr'])
vocab_categories = vect2.get_feature_names_out()
text_transformed_2 = vect2.transform(df['categoriesStr'])

In [53]:
len(vocab_categories)

701

In [54]:
vocab_categories

array(['22', '30', 'added', 'advance', 'alabama', 'alaska', 'alcoholic',
       'almond', 'amaretto', 'anchovy', 'angeles', 'anise', 'anniversary',
       'anthony', 'aperitif', 'app', 'appetizer', 'apple', 'appétit',
       'apricot', 'arizona', 'armagnac', 'artichoke', 'arugula', 'asian',
       'asparagus', 'aspen', 'atlanta', 'australia', 'avocado',
       'backyard', 'bacon', 'bake', 'banana', 'barbecue', 'barley',
       'basil', 'bass', 'bastille', 'bbq', 'beach', 'bean', 'beef',
       'beer', 'beet', 'bell', 'berry', 'beverly', 'birthday', 'biscuit',
       'bitters', 'blackberry', 'blender', 'blue', 'blueberry', 'boil',
       'boiler', 'bok', 'bon', 'boston', 'bourbon', 'bourdain', 'bowl',
       'braise', 'bran', 'brandy', 'bread', 'breadcrumbs', 'breakfast',
       'brie', 'brine', 'brisket', 'broccoli', 'broil', 'brooklyn',
       'brown', 'brownie', 'brunch', 'brussel', 'buffalo', 'buffet',
       'bulgaria', 'bulgur', 'burrito', 'butter', 'buttermilk',
       'butternut

In [44]:
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer

# vectorizer = TfidfVectorizer(stop_words = "english", min_df=2)
# vectorizer = TfidfVectorizer(stop_words = "english")
# vectorizer = CountVectorizer(stop_words = "english")
vectorizer = CountVectorizer(tokenizer=my_tokenizer,
                       min_df=5)

TF_IDF_matrix = vectorizer.fit_transform(df['ingredientsStr'])

c:\Users\anami\anaconda3\Lib\site-packages\sklearn\feature_extraction\text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


In [45]:
TF_IDF_matrix.shape

(14526, 2168)

In [46]:
TF_IDF_matrix

<14526x2168 sparse matrix of type '<class 'numpy.int64'>'
	with 344380 stored elements in Compressed Sparse Row format>

In [47]:
# Artichoke and Parmesan Risotto
df[df['title'].str.contains('Chicken Parmesan', na=False)]

recipeId  calories  rating                         title  \
4918       6368    1842.0   5.000              Chicken Parmesan   
7482       9873    3392.0   3.750        Chicken Parmesan Heros   
9982      13372     610.0   4.375          New Chicken Parmesan   
12825     17557     917.0   5.000  Quick Baked Chicken Parmesan   

                                           directionsStr  \
4918   ['Place breadcrumbs and flour in 2 separate sh...   
7482   ['Heat olive oil in a 4- to 5-quart heavy sauc...   
9982   ['Preheat oven to 500° F. Whisk first 3 ingred...   
12825  ['Arrange racks in top and bottom of oven and ...   

                                           categoriesStr  \
4918   ['Chicken', 'Tomato', 'Broil', 'Kid-Friendly',...   
7482   ['Sandwich', 'Cheese', 'Chicken', 'Poultry', '...   
9982   ['Chicken', 'Tomato', 'Roast', 'Kid-Friendly',...   
12825  ['22-Minute Meals', 'Chicken', 'Parmesan', 'To...   

                                          ingredientsStr  
4918   ['2 cups fine dry breadcrumbs', '1 cup all-pur...  
7482   ['3 tablespoons olive oil', '1 small onion, fi...  
9982   ['1/3 cup extra-virgin olive oil', '2 large ga...  
12825  ['2 large eggs', '1 1/2 cups breadcrumbs or pa...

In [48]:
df[df['title'].str.contains('Chicken', na=False)]

recipeId  calories  rating  \
25           35     625.0   3.750   
37           53    1203.0   5.000   
61           80    1172.0   4.375   
63           82     682.0   4.375   
67           87    1143.0   0.000   
...         ...       ...     ...   
14445     20007     936.0   4.375   
14473     20050     356.0   3.750   
14475     20052     878.0   4.375   
14497     20091    1096.0   3.750   
14518     20121    1086.0   5.000   

                                                   title  \
25                                         Aztec Chicken   
37           Pancetta Roast Chicken with Walnut Stuffing   
61     Braised Chicken and Rice with Orange, Saffron,...   
63                   Chicken in Green Pumpkin-Seed Sauce   
67                 Roast Chicken With Sorghum and Squash   
...                                                  ...   
14445  Braised Chicken with Smoked Ham, Chestnuts, an...   
14473                                  Hot Chicken Salad   
14475                                 Chicken Tetrazzini   
14497                     Chicken with Raisins and Lemon   
14518  Crispy Salt and Pepper Chicken with Caramelize...   

                                           directionsStr  \
25     ['Melt 2 tablespoons butter with vegetable oil...   
37     ['Preheat oven to 400°F. Melt 1/4 cup butter i...   
61     ['Rinse the rice in a sieve under cold running...   
63     ['Bring all ingredients to boil in large pot. ...   
67     ['Bring 5 cups water to a boil in a medium pot...   
...                                                  ...   
14445  ['Bring water to a simmer in a small saucepan,...   
14473  ['1. Preheat the oven to 375°F. Spray a 13-by-...   
14475  ['Bring chicken bones, broth, carrot, onion, c...   
14497  ['Arrange chicken in single layer in large Dut...   
14518  ['1. Preheat the oven to 400°F. Spray a baking...   

                                           categoriesStr  \
25     ['Chicken', 'Olive', 'Onion', 'Sauté', 'Dinner...   
37     ['Chicken', 'Roast', 'High Fiber', 'Dinner', '...   
61     ['Chicken', 'Citrus', 'Fruit', 'Nut', 'Poultry...   
63     ['Chicken', 'Low/No Sugar', 'Cinco de Mayo', '...   
67     ['Bon Appétit', 'Dinner', 'Chicken', 'Grains',...   
...                                                  ...   
14445  ['Chicken', 'Ginger', 'Braise', 'Marinate', 'D...   
14473  ['Cheese', 'Chicken', 'Nut', 'Poultry', 'Bake'...   
14475  ['Chicken', 'Mushroom', 'Pasta', 'Bake', 'Supe...   
14497  ['Chicken', 'Potato', 'Poultry', 'Lemon', 'Rai...   
14518  ['Chicken', 'Poultry', 'Bake', 'Passover', 'Fe...   

                                          ingredientsStr  
25     ['6 tablespoons (3/4 stick) chilled butter', '...  
37     ['8 tablespoons (1 stick) butter, divided', 'C...  
61     ['1 1/2 cups brown basmati rice', '1/4 cup oli...  
63     ['5 cups water', '6 chicken thighs with skin a...  
67     ['Kosher salt', '1 cup sorghum', '1/2 large bu...  
...                                                  ...  
14445  ['2 3/4 cups water', '12 dried Chinese black m...  
14473  ['2 cups cooked chicken breast meat, cubed (Yo...  
14475  ['1 to 1 1/2 pound chicken bones (from 2 cooke...  
14497  ['1 3 1/2-pound chicken, cut into 8 pieces', '...  
14518  ['Cooking spray', 'One 3 1/2-pound chicken, cu...  

[919 rows x 7 columns]

In [49]:
TF_IDF_matrix[(df['title'] == 'Chicken Parmesan').values].todense().squeeze()

matrix([[0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [50]:
from sklearn.metrics.pairwise import cosine_similarity

recipe_1 = TF_IDF_matrix[(df['title'] == 'Chicken Parmesan').values,]
recipe_2 = TF_IDF_matrix[(df['title'] == 'Chicken Parmesan Heros').values,]

print("Similarity:", cosine_similarity(recipe_1, recipe_2)) # Notice the result is a 2D 1X1 array, so to grab
                                                          # the number we will need to index

Similarity: [[0.34776522]]


In [51]:
recipe_3 = TF_IDF_matrix[(df['title'] == 'Artichoke and Parmesan Risotto').values,]
print("Similarity:", cosine_similarity(recipe_1, recipe_3))

Similarity: [[0.20628425]]


TODO - Ana - Look at the actual recipes and see whether the similarity makes sense

In [52]:
from sklearn.metrics.pairwise import cosine_similarity
similarities = cosine_similarity(TF_IDF_matrix, dense_output=False)

In [53]:
# Check the shape
# rows and columns should be equal, and the number of movies we started with (rows)
similarities.shape

(14526, 14526)

In [54]:
# Test with a sample recipe
df[df['title'] == 'Chicken Parmesan']
# print(df.loc[df['recipeId'] == 6368, ['title']].values[0])

recipeId  calories  rating             title  \
4918      6368    1842.0     5.0  Chicken Parmesan   

                                          directionsStr  \
4918  ['Place breadcrumbs and flour in 2 separate sh...   

                                          categoriesStr  \
4918  ['Chicken', 'Tomato', 'Broil', 'Kid-Friendly',...   

                                         ingredientsStr  
4918  ['2 cups fine dry breadcrumbs', '1 cup all-pur...

In [55]:
# Get the column based upon the index
recipe_index = df[df['title'] == 'Chicken Parmesan'].index

# Create a dataframe with the movie titles
sim_df = pd.DataFrame({'recipe': df['title'],
                       'similarity': np.array(similarities[recipe_index, :].todense()).squeeze()})

In [56]:
# Return the top 10 most similar recipes
sim_df.sort_values(by='similarity', ascending=False).head(10)

recipe  similarity
4918                         Chicken Parmesan    1.000000
12825            Quick Baked Chicken Parmesan    0.553191
6860     Chicken and Dumplings with Mushrooms    0.522931
11718                  Fried Chicken Biscuits    0.520939
13533  Fusilli with Shrimp and Paneed Chicken    0.517769
1937              East-West Barbecued Chicken    0.506168
8877       Pepper, Rosemary, and Cheese Bread    0.505992
7471   Fried Chicken Thighs with Cheesy Grits    0.503038
4490                         Parmesan Muffins    0.501956
14180             BA's Best Eggplant Parmesan    0.500428

With TFIDF Vectorizer:

| recipe |                                       similarity |          |
|-------:|-------------------------------------------------:|----------|
|   4918 |                                 Chicken Parmesan | 1.000000 |
|  11718 |                           Fried Chicken Biscuits | 0.424981 |
|  12825 |                     Quick Baked Chicken Parmesan | 0.424046 |
|   1165 |                          Mozzarella Pesto Spread | 0.412968 |
|   9877 |         Rigatoni with Cheese and Italian Sausage | 0.360720 |
|  14180 |                      BA's Best Eggplant Parmesan | 0.356189 |
|   5786 |                         Horseradish-Yogurt Sauce | 0.351933 |
|   3927 | Chunky Two-Cheese Potatoes with Garlic and Pesto | 0.333565 |
|  13533 |           Fusilli with Shrimp and Paneed Chicken | 0.331478 |
|   3123 |    Spicy Lamb Pizza With Parsley–Red Onion Salad | 0.328433 |

Results with CountVectorizer seem better:

|       |                                            recipe | similarity |
|------:|--------------------------------------------------:|------------|
|  4918 |                                  Chicken Parmesan |   1.000000 |
| 13533 |            Fusilli with Shrimp and Paneed Chicken |   0.587137 |
| 12480 |                          Spicy Oven-Fried Chicken |   0.574641 |
| 11718 |                            Fried Chicken Biscuits |   0.559149 |
| 14180 |                       BA's Best Eggplant Parmesan |   0.557207 |
|  7749 |            Lamb and Eggplant Casserole (Moussaka) |   0.556499 |
| 11245 |                                             Pinon |   0.548443 |
|  9785 | Breaded Skinless Fish Fillets with Red Pepper ... |   0.540222 |
|  6860 |              Chicken and Dumplings with Mushrooms |   0.539425 |
| 11336 | Crispy Chicken Sandwich with Buttermilk Slaw a... |   0.539164 |

TODO: Ana check these results - something does not make sense, should I try with CountVectorizer? Still getting some weird results like Pinon.

Tried with custom tokenizer that removed most measurements and adjectives - TODO see if there is a library for this e.g. https://stackoverflow.com/questions/33587667/extracting-all-nouns-from-a-text-file-using-nltk

The results are better now:
|   |       |                                 recipe | similarity |
|---|------:|---------------------------------------:|------------|
|   |  4918 |                       Chicken Parmesan |   1.000000 |
|   | 12825 |           Quick Baked Chicken Parmesan |   0.553191 |
|   |  6860 |   Chicken and Dumplings with Mushrooms |   0.522931 |
|   | 11718 |                 Fried Chicken Biscuits |   0.520939 |
|   | 13533 | Fusilli with Shrimp and Paneed Chicken |   0.517769 |
|   |  1937 |            East-West Barbecued Chicken |   0.506168 |
|   |  8877 |     Pepper, Rosemary, and Cheese Bread |   0.505992 |
|   |  7471 | Fried Chicken Thighs with Cheesy Grits |   0.503038 |
|   |  4490 |                       Parmesan Muffins |   0.501956 |
|   | 14180 |            BA's Best Eggplant Parmesan |   0.500428 |

In [91]:
selector = (df['title'] == 'Chicken Parmesan') | (df['title'] == 'Fried Chicken Biscuits') | (df['title'] == 'Pinon')
df.loc[selector,['title', 'ingredientsStr']].values

array([['Chicken Parmesan',
        '[\'2 cups fine dry breadcrumbs\', \'1 cup all-purpose flour\', \'4 large eggs\', \'1 cup whole milk\', \'8 small skinless, boneless chicken thighs, pounded to 1/2" thickness\', \'Kosher salt, freshly ground pepper\', \'N/A freshly ground pepper\', \'8 tablespoons olive oil\', \'8 tablespoons prepared sun-dried tomato pesto\', \'1 pound fresh mozzarella, cut into 8 slices\', \'1/2 teaspoon crushed red pepper flakes\', \'4 cups prepared marinara sauce, warmed\', \'Finely grated Parmesan (for serving)\']'],
       ['Pinon',
        "['1 medium onion', '1/2 small green bell pepper', '1/2 small red bell pepper', 'a 14- to 16-ounce can whole tomatoes', '1/3 cup drained pimiento-stuffed green olives', '1 pound ground beef chuck', '1/4 teaspoon salt', '1/4 teaspoon freshly ground black pepper', '1/2 cup tomato sauce', '2 tablespoons raisins', '1 tablespoon cider vinegar', '2 bay leaves', '1/4 teaspoon ground achiote (optional)', '6 semi-ripe (yellow with so

[['Chicken Parmesan',
         breadcrumbs\flour\eggs\'milk\'chicken thighs, \'Kosher salt, freshly ground pepper\'olive oil\'tomato pesto\', mozzarella, red pepper flakes\'marinara sauce, warmed\', \'Finely grated Parmesan (for serving)\']'],
       ['Fried Chicken Biscuits',
        buttermilk\', \'1 tablespoon hot sauce\', \'1/2 teaspoon kosher salt\', \'1/4 teaspoon freshly ground black pepper\', \'1/4 teaspoon cayenne pepper\', \'4 skinless, boneless chicken cutlets (about 1 1/2 pounds total), pounded to 1/4" thickness\', \'1/2 cup (1 stick) unsalted butter, room temperature\', \'1/4 cup honey\', \'1/2 teaspoon kosher salt, plus more\', \'Vegetable oil (for frying; about 2 1/2 cups)\', \'2 cups panko (Japanese breadcrumbs)\', \'Freshly ground black pepper\', \'4 Pies-N-Thighs Biscuits\', \'Hot sauce (for serving)\']']],
      dtype=object)

In [57]:
# Check for Aritchoke
# Get the column based upon the index
recipe_index_2 = df[df['title'] == 'Artichoke and Parmesan Risotto'].index

# Create a dataframe with the movie titles
sim_df_2 = pd.DataFrame({'recipe': df['title'],
                       'similarity': np.array(similarities[recipe_index_2, :].todense()).squeeze()})

# Return the top 10 most similar recipes
sim_df_2.sort_values(by='similarity', ascending=False).head(10)

recipe  similarity
14522                     Artichoke and Parmesan Risotto    1.000000
6459                                   Asparagus Risotto    0.757240
10936                   Risotto with Squash and Pancetta    0.700301
5060          Shrimp Risotto with Baby Spinach and Basil    0.699400
4857                                   Radicchio Risotto    0.659966
10112                           Porcini Mushroom Risotto    0.646762
11204         Asparagus and Leek Risotto with Prosciutto    0.625779
5028                       Celery Root Risotto and Pesto    0.625463
11013                                    Spinach Risotto    0.625463
8493   Butternut Squash, Rosemary, and Blue Cheese Ri...    0.625000

In [36]:
import nltk

lines = 'lines is some string of words'
# function to test if something is a noun
is_noun = lambda pos: pos[:2] == 'NN'
# do the nlp stuff
tokenized = nltk.word_tokenize(lines)
nouns = [word for (word, pos) in nltk.pos_tag(tokenized) if is_noun(pos)]

print(nouns)

['lines', 'string', 'words']
